### PR TITLE
C API: Add libflake-c

### LINKED ISSUE #11399: `getFlake` not available through C API
**Describe the bug**

- `libnixflake` has been factored out since #9063
- global state doesn't solve it by just linking since #11007
  - we don't have a `libnixflake-c` to explicitly init the lib

**Steps To Reproduce**

1. evaluate `builtins.getFlake` using the C API

**Expected behavior**

When the user config allows `flakes` to be used, `getFlake` is available.

**`nix-env --version` output**

**Additional context**

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

### PR DESCRIPTION
**Pull request by roberth**
## Motivation

Nix 2.24's C API is currently incapable of providing flake functionality, because `libflake` wasn't wrapped before https://github.com/NixOS/nix/pull/9063

This PR provides a very basic `libflake` that restores `builtins.getFlake`. (a feature compensating for a regression)

## Context

- #ZurichZHF
- Fixes #11399
- https://github.com/NixOS/nix/pull/9063

---

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: src/libexpr-tests/nix_api_expr.cc [hunk 1]
```diff
@@ -7,12 +7,49 @@
...
+    auto nixos = tmpDir + "/cfg";
+    std::filesystem::create_directories(nixpkgs);
+    std::filesystem::create_directories(nixos);
+
+    std::string nixpkgsEntry = "nixpkgs=" + nixpkgs;
+    std::string nixosEntry = "nixos-config=" + nixos;
+    const char * lookupPath[] = {nixpkgsEntry.c_str(), nixosEntry.c_str(), nullptr};
+
+    auto builder = nix_eval_state_builder_new(ctx, store);
+    assert_ctx_ok();
```
Comment (Mic92) [inline src/libexpr-tests/nix_api_expr.cc:31]: IF these asserts trigger are is the temporary directory still deleted i.e. is this unwinding the stack rather?

Comment (roberth) [reply src/libexpr-tests/nix_api_expr.cc:31]: It did unwind properly, but it urns out `FAIL()` creates a `return;`, which makes this helper ineffective. Fixed in f0e4a81, _test: Change fail to throw_

File: src/libexpr-c/nix_api_expr.cc [hunk 2]
```diff
@@ -14,10 +12,37 @@
 #include "nix_api_util.h"
 #include "nix_api_util_internal.h"
 
+#include <cstring>
+#include <stdexcept>
+#include <string>
```
Comment (Ericson2314) [inline src/libexpr-c/nix_api_expr.cc:17]: I think putting the external ones at the top is better? It's the rest of our code that should be changed to do this.

**Review by Ericson2314 — APPROVED** [review]

File: src/libexpr-c/nix_api_expr.cc (hunk 2 above)
Comment (roberth) [reply src/libexpr-c/nix_api_expr.cc]: :facepalm:
Done.

**Pull request merged**